samples version of the game engine now in tutorial.

gpx1000 opened the pull request:
NB: if you have the bistro from before, please grab the latest to be able to see the animations of the ceiling fan.


Apply consistent formatting, improve code readability, and add/update copyright headers across multiple files.

This is the full simple game engine with all features implemented including: 

- streaming textures in separate threads with transfer queue ray query rendeirng pipeline
- forward+ rendering pipeline
- reflections
- thick and thin glass
- translucency
- better lighting fixes so light doesn't appear over bright animation from GLTF
- robustness2
- synchronization2
- Dynamic rendering local read
- shader tile image
- culling

Incorporated PRs for:

- Android support
- Windows support
- memory leak fixes

gpx1000 commented:
Yep, I'm working those into the tutorial itself.  Should have them soon

SaschaWillems commented:
If I increase the watchdog timeout, same issue as I had with an earlier version: In debug it starts loading things and at some point just gets stuck, and nothing ever happens.

In release it outright crashes in the driver right at startup (Nvidia RTX 4070).

gpx1000 commented:
I fixed the cmake by committing the cmake folder that was missing.

The watchdog I might remove; I don't know if it makes sense for slower computers.  2 seconds between update should have been fine.  Maybe 10 seconds?  Or maybe just remove it.

Lemme see if I can find all the places where I messed up the scoping.

The indention, was done by the samples' script.  I'll redo it and make it cleaner.  Just about to add the updates to the tutorial text.

marty-johnson59 commented:
Hmm, might be cleaner to have the same license.  Is that possible @gpx1000?  If not, I can check with Khronos legal to see if mixing licenses is an issue..

gpx1000 commented:
@SaschaWillems I *think* I hopefully got all of the potential places that windows could have crash potential problems.  I also did a bunch of work on improving the frame rate.  I now see 60fps in Linux for all render types while in debug build, except for forward+ (for hopefully obvious reasons (i.e. if you want boundless dynamic lighting capabilities without deferred there's significant overhead to that kind of trade off; could refine this further to get frame rate up there, but these optimizations are the ones I can do in a day).
i'm on a NVIDIA 2090 6 year old laptop.
Lemme know if you run into any more issues.  I'll get a Windows machine to validate these on if you run into problems there.

SaschaWillems commented:
No need to hurry ;)

8SecSleeper commented:
I disabled frustum culling, lod functions, I'm seeing roughly 50 percent drop in FPS vs the previous version that's been committed, Even though it says ray query's not supported so I assume they aren't being ran.

Also there's CMake issues for several things. Which needed to be changed, I used the same CMake profiles I made for the committed version, Since the new versions being pushed didn't work and are missing 2 files still. I suggest testing this on a fresh windows install to see what's missing.

gpx1000 commented:
Thanks for the report.  Could you be more specific about what's missing for you in the CMake that's been committed?  I am on a local fresh windows install and I haven't run into issues with the CMake.  Are you using vcpkg?   I assume you're on windows because you're saying you're seeing a 50 % frame drop.  In linux I get 60fps with all the various options.
I locally have a fix for the ray queries not working that I'm going to commit as soon as I get a full fix for the hang on start issue.  That's taking some refactoring so it might not land tonight; but maybe in the next few days.

gpx1000 commented:
I should do a quick version that adds to the CI, not sure if it already checks the game engine or not (can't remember).

gpx1000 commented:
I went ahead and added ci to this last night.  Need to fix Linux again.   But hopefully that'll help as well. 

Thanks for testing.  I expect there's plenty of bugs in the pr.  Spaces in the path and making sure the path is under a certain length are always the fun ways windows throws is curve balls.  I'll have to think about where and how best to document that as it could appear anywhere there's file interaction and paths.

gpx1000 commented:
Ahh... okay, getting CI working required some of that.  I'll nuke that from the vcpkg in the next commit.  Still working on Linux.

gpx1000 commented:
I've added shadows, and reworked quite a bit of things on how the lighting works.  Lots of optimizations.  Shadows are now handled by Ray Query in all pipelines.
Additionally, I added support for caching the simple game engine builds in CI.  On my system I'm getting 80+FPS across all types of rendering.  Forward+ is still slower, but that's part of the Forward+ algorithm.  With the optimizations and the reductions in duplicated work, this should now be pretty fast in Windows. 
Oh, I also added a chapter on shadows and explained how they work in the engine along with a generic (gentle) introduction to what shadows are and how to get more resources.

gpx1000 commented:
Testing in Windows shows everything is still working, but the frame rate is much lower than in Linux.  Investigating.

gpx1000 commented:
I was missing some of my Windows changes.  After finally getting everything merged with what was on the windows side, I went through and had to fix the decals; and some other bugs.  It now renders clean for me and I think I addressed all of your concerns above.  However, I never saw any warnings surrounding locks.  lemme know if you still see those warnings and where they are and I'll take a look.

SaschaWillems commented:
Somehow forgot to save my comment: This does work fine for me on Windows, both in debug and release, looks good and runs good. Remaining things are just minor and can be easily fixed afterwards :)
